packages/expression-parser/src/expression-parser/ExpressionParser.js

```diff
@@ -0,0 +1,110 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+import { create, all } from 'mathjs';
+
+export class ExpressionParser {
+  constructor() {
+    this.math = create(all, {});
+    this.parser = this.math.parser();
+  }
+
+  /**
+   * Evaluate an expression.
+   * Before evaluating an expression, the scope will be sanitized to fix up any invalid variables that we want to support
+   * @param {*} expression
+   */
+  evaluate(expression) {
+    const originalScope = this.parser.getAll();
+
+    // Sanitize scope and expression to fix any invalid variables before evaluating the expression
+    const sanitizedExpression = this.sanitizeExpression(expression, originalScope);
+
+    // Evaluate the sanitized expression
+    const result = this.parser.evaluate(sanitizedExpression);
+
+    // Reset the scope variables to original;
+    this.clearScope();
+    this.setScope(originalScope);
+
+    return result;
+  }
+
+  /**
+   * Set the parser's scope.
+   * @param {*} scope
+   * @param {*} defaultScope
+   */
+  setScope(scope = {}, defaultScope = {}) {
+    Object.entries(scope).forEach(([name, value]) => {
+      const expressionValue = value || defaultScope[name] || 0;
+      this.set(name, expressionValue);
```

Comment:
```suggestion
      this.parser.set(name, expressionValue);
```
And delete the one below - doesn't seem worth its own function to me?

Reply:
I was thinking so that when setting a scope value outside of this package, we don't have to do something like `expressionParser.parser.set('a', 5)`. But it's very trivial...

Reply:
Oh true. Do you think that will actually be used? I would have imagined the use cases to either be one of the following, rather than setting scope a field at a time:
````
const scope = { a: 1, b: 2 };
const expression1 = 'a + b';
const expression2 = 'a x b';

// option 1
const parser = new ExpressionParser(scope);
parser.evaluate(expression1);
parser.evaluate(expression2);

// option 2
const parser = new ExpressionParser();
parser.setScope(scope);
parser.evaluate(expression1);
parser.evaluate(expression2);

// option 3
const parser = new ExpressionParser();
parser.evaluate(expression1, scope); // order may be reversed
parser.evaluate(expression2, scope);
```